### PR TITLE
[move-prover] Added Boogie model for serialize to prelude.bpl

### PR DESCRIPTION
**Pull request by DavidLDill**
Move datastructures are converted to vector<u8> for sha2/3, crypto and other purposes by a native library called "serialize".  There is no "deserialize" in Move and no plans to create one.

The new Boogie code for serialize maps a Move value to an arbitrary non-empty vector<8>.  There is a an additional axiom that constrains serialize to be an injection, because (I believe) the correctness of Move modules will sometimes depend on this property. However, the model is a conservative approximation of the actual function because it does not guarantee that the bytes in the serialized value correspond to what the native serialize function does, or even that the length of the result is the same as the actual serialize.

We should be able to verify properties that require serialize to be an injection, but won't be able to verify code that accesses and manipulates bytes from a serialized value or, for example, code that depends on the length of serialized values.

There are test cases serialize-model.move and serialize-model-invalid.move, also. Some new .exp files were created for stdlib files that use lcs.move.

## Motivation

(Write your motivation for proposed changes here.)

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

(Write your answer here.)

## Test Plan

(Share your test plan here. If you changed code, please provide us with clear instructions for verifying that your changes work.)

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/libra/website, and link to your PR here.)

### CODE REVIEW

File: language/move-prover/src/prelude.bpl [hunk 1]
```diff
@@ -541,7 +541,7 @@ procedure {:inline 1} BorrowField(src: Reference, f: FieldName) returns (dst: Re
 
     p := p#Reference(src);
     size := size#Path(p);
-	p := Path(p#Path(p)[size := f], size+1);
+        p := Path(p#Path(p)[size := f], size+1);
```
Comment (wrwg) [inline language/move-prover/src/prelude.bpl:544]: Is this perhaps a tab character? If so please avoid tabs because they are showing differently here and there. Most editors should have settings to have the same benefits without the tab character,

File: language/move-prover/src/prelude.bpl [hunk 2]
```diff
@@ -1131,6 +1111,40 @@ procedure {:inline 1} Signature_ed25519_threshold_verify(bitmap: Value, signatur
...
+
+// Serialize is modeled as (almost) uninterpreted function, with an additional
+// axiom to say it's an injection.
+
+function $serialize(ta: TypeValue, v: Value): Value;
+
+// This says that $serialize respects isEquals (substitution property)
+// Without this, Boogie will get false positives where v1, v2 differ at invalid
+// indices.
+axiom (forall ta: TypeValue ::
```
Comment (wrwg) [inline language/move-prover/src/prelude.bpl:1127]: Just wondering: does this need to be a nested quantor?

File: language/move-prover/src/prelude.bpl [hunk 3]
```diff
@@ -1131,6 +1111,40 @@ procedure {:inline 1} Signature_ed25519_threshold_verify(bitmap: Value, signatur
...
+axiom (forall ta1, ta2: TypeValue ::
+       (forall v1, v2: Value :: IsEqual($serialize(ta1, v1), $serialize(ta2, v2))
+           ==> IsEqual(v1, v2) && (ta1 == ta2)));
+
+
+procedure $LCS_to_bytes(ta: TypeValue, r: Reference) returns (res: Value);
+ensures res == $serialize(ta, $Dereference($m, r));
+ensures $IsValidU8Vector(res);    // result is a legal vector of U8s.
+ensures $vlen(res) > 0;
+ensures is#Boolean($Dereference($m, r)) ==> $vlen(res) == 1;
```
Comment (wrwg) [inline language/move-prover/src/prelude.bpl:1141]: Why is this needed? Also, if we do it for bool, why not for others? (Could not understand what you mean with "wrong" below.)

File: language/move-prover/tests/sources/serialize_model.move [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    }
+    spec fun lcs_test1 {
+        aborts_if false;
+        ensures result_1 == result_2 ==> v1 == v2;
+        ensures v1 == v2 ==> result_1 == result_2;
+        // it knows result is vector<u8>
+        ensures len(result_1) > 0 ==> result_1[0] <= max_u8();
+    }
+
+    // FIXME: Commented out because prelude "ensures" about lengths are wrong.
```
Comment (wrwg) [inline language/move-prover/tests/sources/serialize_model.move:25]: Is there an issue or a description somewhere what is wrong?

File: language/move-prover/src/prelude.bpl (hunk 2 above)
Comment (DavidLDill) [reply language/move-prover/src/prelude.bpl:1127]: I just did that because I'm not sure how general Boogie quantifier syntax is.  I'm guessing that it doesn't matter to Boogie, and it seems to work in practice.

File: language/move-prover/src/prelude.bpl (hunk 3 above)
Comment (DavidLDill) [reply language/move-prover/src/prelude.bpl:1141]: We know the lengths of some primitives.  I don't know if it matters, but my guess is that it has low cost for Boogie/Z3 and that it may help sometimes.  One scenario where it might make a difference is when concatenating serialized values.  If  x != z, you can conclude that x*y != z*w IF len(x) == len(z).  That may be important at some point (guid in libra_account concatenates some things and wants them to be unique, but I haven't tried it out to see if this is needed).

File: language/move-prover/tests/sources/serialize_model.move (hunk 4 above)
Comment (DavidLDill) [reply language/move-prover/tests/sources/serialize_model.move:25]: I will probably file one, but it wasn't high priority.  This is related to my answer to. your question about the length of bool.  I thought we knew the lengths of U8, U64, U128, but we don't really know the types.  They're all Integer, and $IsValidU8 etc. just check the bounds on the integer, which may not be the same as the length of the data encoding.  We need a little more bookkeeping to keep track of the length in bytes of the actual data.  I left this in so I wouldn't have to regenerate it if and when we fix that.

File: language/move-prover/src/prelude.bpl (hunk 3 above)
Comment (wrwg) [reply language/move-prover/src/prelude.bpl:1141]: I still don't understand why you have then commented out for the other primitives, and what you mean with "wrong". 

Also someone looking at this in the future, might not know why you added this, and gets confused somehow thinking this is a semantic requirement. If you keep it please add a comment why its there.

File: language/move-prover/tests/sources/serialize_model.move (hunk 4 above)
Comment (wrwg) [reply language/move-prover/tests/sources/serialize_model.move:25]: Ping on this comment. I just don't understand what is "wrong" here and why you left a "FIXME".

File: language/move-prover/src/prelude.bpl (hunk 3 above)
Comment (wrwg) [reply language/move-prover/src/prelude.bpl:1141]: Can we perhaps just leave this length out as we are right now have just a guess that it might be helpful, and it anyway doesn't work for anything else than bool? I find it highly confusing as it is now.
